### PR TITLE
Add Service Account authentication for HTTP Source Plugins

### PR DESCRIPTION
**Pull request by VictorD-Veolia**
This PR adds the capability to use Service Account as an authentication method for HTTP Source Plugins.

### CODE REVIEW

File: src/main/java/io/cdap/plugin/http/source/common/BaseHttpSourceConfig.java [hunk 1]
```diff
@@ -794,6 +827,13 @@ PAGINATION_INDEX_PLACEHOLDER, getPaginationType()),
       assertIsSet(getRefreshToken(), PROPERTY_REFRESH_TOKEN, reasonOauth2);
     }
 
+    // Validate service account auth properties
+    if (!containsMacro(PROPERTY_SERVICE_ACCOUNT_PRIVATE_KEY_JSON) && this.getServiceAccountEnabled()) {
+      String reasonSA = "Service Account Authentication is enabled";
+      //assertIsSet(getServiceAccountScope(), PROPERTY_SERVICE_ACCOUNT_SCOPE, reasonSA);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/http/source/common/BaseHttpSourceConfig.java:833]: remove commented out line

File: src/main/java/io/cdap/plugin/http/source/common/http/ServiceAccountTokenHandler.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/http/source/common/http/ServiceAccountTokenHandler.java:2]: 2022

File: src/main/java/io/cdap/plugin/http/source/common/http/ServiceAccountTokenHandler.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        } else {
+            if (tokenExpireSoon()) {
+                token = googleCredentials.refreshAccessToken();
+            }
+        }
+
+        if (token != null) {
+            return token;
+        } else {
+            throw new AuthenticationException("Tried to get access token from GoogleCrendentials but got " + token);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/http/source/common/http/ServiceAccountTokenHandler.java:78]: It's always going to be "but got null". Change error msg to something like "Failed to get access token"

File: src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java [hunk 4]
```diff
@@ -59,29 +65,63 @@ public HttpClient(BaseHttpSourceConfig config) {
...
     }
+
+    if (config.getServiceAccountEnabled()) { // Service account authent used
+      try {
+        serviceAccountTokenHandler = new ServiceAccountTokenHandler(
+                config.getServiceAccountScope(),
+                config.getServiceAccountPrivateKeyJson()
+        );
+      } catch (Exception e) {
+        LOG.error("Error while getting Service Account Access token " + e);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java:76]: We shouldn't ignore exception here.

File: src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java [hunk 5]
```diff
@@ -59,29 +65,63 @@ public HttpClient(BaseHttpSourceConfig config) {
...
     if (requestBody != null) {
       request.setEntity(requestBody);
     }
 
-    return httpClient.execute(request);
+    CloseableHttpResponse response = httpClient.execute(request);
+
+    if (retries > 0 && config.getServiceAccountEnabled()) {
+      HttpResponse httpResponse = new HttpResponse(response);
+      if (httpResponse.getStatusCode() == 401) { // If 'Unhauthorized' error, recreate client with fresh token
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java:118]: I don't think this retry logic is needed as we're recreating the client with a fresh token if it's going to expire soon (line 101)

**Review by rmstar — COMMENTED** [review]
- Please create a Jira for this feature and link it in this PR
- Please fix merge conflicts
- Consider adding a unit test

**Comment by VictorD-Veolia** [comment]
The service account authentication has been implemented in the meantime by @sumengwang  : https://github.com/data-integrations/http/commit/d4ba0e4500167da784e721d39376a66e3ed8f978

This pull request has no reason to to be merged anymore

**Pull request closed without merging**